portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -23,9 +32,48 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Kyle!</h1>");
+    int numComments = Integer.parseInt(getParameter(request, "num-comments", "5"));
+    int page = Integer.parseInt(getParameter(request, "page", "1"));
+    String other = getParameter(request, "other", "None");
+    Gson gson = new Gson();
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    List<Entity> results = datastore.prepare(query).asList(
+        FetchOptions.Builder.withLimit(numComments).offset(numComments * (page - 1)));
+    ArrayList<String> comments = new ArrayList<String>();
+    for (Entity entity : results) {
+      comments.add((String) entity.getProperty("text"));
+    }
+    String json = gson.toJson(comments);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long timestamp = System.currentTimeMillis();
+    String text = getParameter(request, "comment-text", "");
+    System.out.println("Add Comment: " + text);
+    if (text.length() > 0) {
+      Entity commentEntity = new Entity("Comment");
+      commentEntity.setProperty("text", text);
+      commentEntity.setProperty("timestamp", timestamp);
+      datastore.put(commentEntity);
+    }
+    response.sendRedirect("/#comments");
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
```

Comment:
maybe rename this function to a meaningful name, e.g. getRequestParameterOrDefault